index.js

```diff
@@ -516,12 +540,39 @@ class KeyringController extends EventEmitter {
    * @param {string} password - The keyring controller password.
    * @returns {Promise<boolean>} Resolves to true once keyrings are persisted.
    */
-  async persistAllKeyrings(password = this.password) {
-    if (typeof password !== 'string') {
+  async persistAllKeyrings(password) {
+    if (password && typeof password !== 'string') {
       throw new Error('KeyringController - password is not a string');
     }
 
-    this.password = password;
+    // MV3: Since we also allow persisting without a password,
+    // we should require this.encryptedKey
+    if (password === undefined && this.encryptedKey === undefined) {
+      return Promise.reject(
+        new Error(
+          'KeyringController - a password or encryptedKey must exist to persist keyrings',
+        ),
+      );
+    }
+
+    let salt = null;
+    if (password) {
+      // MV3: If this is a migration or new password-driven login, we should
+      // create or rotate the salt
+      salt = encryptor.generateSalt();
+
+      // MV3: Since there's a new salt, we need to generate a new encrypted key
+      // for use in the
+      this.encryptedKey = await this._generateEncryptedKey(password, salt);
```

Comment:
We're overriding the encryptedKey field before going through the (**asynchronous**) process of preparing the new vault.
That leaves a possibility (because of bad timing or an error) for the key in this.encryptedKey to not be a key that can open the current vault.

It's not a big deal, worst case scenario it fails and forces the user to supply password again. Might be annoying if happening right after they supplied the password already.

Keeping the key in local variable and setting it in the same synchronous execution as the call to store the new vault would eliminate that gap.


And while we're at it - where/when is the key being saved to survive a restart? Will this flow save it too early? Why is it not saved in the same synchronous execution as the vault it unlocks?